%Move 1: Background to the Dissertation
%Move 2: Statement of the Problem/Gap in the Research
%Move 3: Purpose of the Dissertation
%Move 4: Work carried out/Methods
%Move 5: Results
%Move 6: Conclusions/Implications
%Move 7: Contribution of the Dissertation 

Alternative splicing is a key means for genetic molecular diviersity
%Alternative splicing is a fundamental part of gene expression 
and its misregulation has been associated with up to 50\% of all known pathogenic mutations, yet the regulatory mechanisms influencing it are still poorly understood. Better understanding of alternative splicing is in part driven by computational models which predict alternative splicing based on genomic sequence information. 

In this work, we show that a dataset which was previously widely used for the quantification of alternative splicing is confounded such that 100 parameters models can match the performance of 20,000 parameter models, calling the meaningfulness of results using this dataset into question. To this end, we construct three new classes of datasets, each based on a different processing method, and show that only one of them provides high enough data quality for our task. We develop a new Deep Learning model which is the first to introduce an attention mechanism to the prediction of alternative splicing events. Evaluating our newly proposed model, we reimplement two models from the literature and find that it outperforms the previous state-of-the-art model for classifying alternative splicing events by 15\%. Finally, we interpret our model and show that it attends to biologically plausible motifs. 

In summary, our work shows that previous results have been based on critically confounded data,
% may not be as meaningful
provides better datasets upon which future work can build and introduces a new state-of-the-art model. 

%maybe: introduce splicing codes as those models which do the thing
%maybe: say that my model is named RASC